clients/src/main/java/org/apache/kafka/common/requests/OffsetCommitRequest.java

```diff
@@ -147,221 +40,115 @@
-                append(", generationId=").append(generationId).
-                append(", offsetData=").append(offsetData).
-                append(")");
-            return bld.toString();
+            return data.toString();
         }
     }
 
-    private OffsetCommitRequest(String groupId, int generationId, String memberId, long retentionTime,
-                                Map<TopicPartition, PartitionData> offsetData, short version) {
+    private final short version;
+
+    public OffsetCommitRequest(OffsetCommitRequestData data, short version) {
         super(ApiKeys.OFFSET_COMMIT, version);
-        this.groupId = groupId;
-        this.generationId = generationId;
-        this.memberId = memberId;
-        this.retentionTime = retentionTime;
-        this.offsetData = offsetData;
+        this.data = data;
+        this.version = version;
     }
 
-    public OffsetCommitRequest(Struct struct, short versionId) {
-        super(ApiKeys.OFFSET_COMMIT, versionId);
-
-        groupId = struct.get(GROUP_ID);
 
-        // These fields only exists in v1.
-        generationId = struct.getOrElse(GENERATION_ID, DEFAULT_GENERATION_ID);
-        memberId = struct.getOrElse(MEMBER_ID, DEFAULT_MEMBER_ID);
+    public OffsetCommitRequest(Struct struct, short version) {
+        super(ApiKeys.OFFSET_COMMIT, version);
+        this.data = new OffsetCommitRequestData(struct, version);
+        this.version = version;
+    }
 
-        // This field only exists in v2
-        retentionTime = struct.getOrElse(RETENTION_TIME, DEFAULT_RETENTION_TIME);
+    public OffsetCommitRequestData data() {
+        return data;
+    }
 
-        offsetData = new HashMap<>();
-        for (Object topicDataObj : struct.get(TOPICS)) {
-            Struct topicData = (Struct) topicDataObj;
-            String topic = topicData.get(TOPIC_NAME);
-            for (Object partitionDataObj : topicData.get(PARTITIONS)) {
-                Struct partitionDataStruct = (Struct) partitionDataObj;
-                int partition = partitionDataStruct.get(PARTITION_ID);
-                long offset = partitionDataStruct.get(COMMITTED_OFFSET);
-                String metadata = partitionDataStruct.get(COMMITTED_METADATA);
-                PartitionData partitionOffset;
-                // This field only exists in v1
-                if (partitionDataStruct.hasField(COMMIT_TIMESTAMP)) {
-                    long timestamp = partitionDataStruct.get(COMMIT_TIMESTAMP);
-                    partitionOffset = new PartitionData(offset, timestamp, metadata);
-                } else {
-                    Optional<Integer> leaderEpochOpt = RequestUtils.getLeaderEpoch(partitionDataStruct,
-                            COMMITTED_LEADER_EPOCH);
-                    partitionOffset = new PartitionData(offset, leaderEpochOpt, metadata);
-                }
-                offsetData.put(new TopicPartition(topic, partition), partitionOffset);
+    public Map<TopicPartition, Long> offsets() {
```

Comment:
Do we really need to create a whole separate map for this?  Seems like we could just use the existing `topics()` map.  Or provide an accessor that just fetched one specific offset, rather than creating a whole map.